tools/aiecc/aiecc.cpp

```diff
@@ -0,0 +1,862 @@
+cl::opt<std::string> XCLBinInstanceName("xclbin-instance-name", cl::desc("Instance name in xclbin metadata"), cl::init("MLIRAIE"));
+cl::opt<std::string> XCLBinKernelID("xclbin-kernel-id", cl::desc("Kernel id in xclbin file"), cl::init("0x901"));
+
+void addLowerToLLVMPasses(OpPassManager &pm) {
+  pm.addPass(createCanonicalizerPass());
+  pm.addPass(createCSEPass());
+  pm.addPass(createConvertVectorToLLVMPass());
+  pm.addPass(memref::createExpandStridedMetadataPass());
+  pm.addPass(createLowerAffinePass());
+  pm.addPass(createConvertMathToLLVMPass());
+  pm.addPass(createArithToLLVMConversionPass());
+  pm.addPass(createFinalizeMemRefToLLVMConversionPass());
+  pm.addPass(createConvertFuncToLLVMPass({.useBarePtrCallConv=true}));
+  pm.addPass(createConvertControlFlowToLLVMPass());
+  pm.addPass(createCanonicalizerPass());
+  pm.addPass(createCSEPass());
+}
+
+struct Toolkit {
+  std::string TargetArch;
+  sfs::path PeanoDir;
+  sfs::path InstallDir;
+  sfs::path AIEToolsDir;
+  sfs::path TempDir;
+};
+
+int runTool(StringRef Program, ArrayRef<StringRef> Args,
+            std::optional<ArrayRef<StringRef>> Env = std::nullopt) {
+  if (Verbose) {
+    llvm::outs() << "Run:";
+    if (Env) {
+      for (auto &s : *Env) {
+        llvm::outs() << " " << s;
+      }
+    }
+    llvm::outs() << " " << Program;
+    for (auto &s : Args) {
+      llvm::outs() << " " << s;
+    }
+    llvm::outs() << "\n";
+  }
+  std::string err_msg;
+  sys::ProcessStatistics stats;
+  std::optional<sys::ProcessStatistics> opt_stats(stats);
+  SmallVector<StringRef, 8> PArgs = {Program};
+  PArgs.append(Args.begin(), Args.end());
+  int result = sys::ExecuteAndWait(Program, PArgs, Env, {}, 0, 0, &err_msg, nullptr, &opt_stats);
+  if (Verbose) {
+    llvm::outs() << (result == 0 ? "Succeeded " : "Failed ") << "in " <<
+      std::chrono::duration_cast<std::chrono::duration<float>>(stats.TotalTime).count() << " code: " << result << "\n";
+  }
+  return result;
+}
+
+template <unsigned N>
+void aieTargetDefines(SmallVector<StringRef, N> &Args, std::string aie_target) {
+  if (aie_target == "AIE2") {
+    Args.push_back("-D__AIEARCH__=20");
+  } else {
+    Args.push_back("-D__AIEARCH__=10");
+  }
+}
+
+// Generate the elf files for the core
+LogicalResult generateCoreElfFiles(ModuleOp moduleOp, sfs::path &objFile, Toolkit &TK) {
+  auto deviceOps = moduleOp.getOps<xilinx::AIE::DeviceOp>();
+  if (!llvm::hasSingleElement(deviceOps)) {
+    return moduleOp.emitOpError("expected a single device op");
+  }
+
+  xilinx::AIE::DeviceOp deviceOp = *deviceOps.begin();
+  auto tileOps = deviceOp.getOps<xilinx::AIE::TileOp>();
+
+  std::string errorMessage;
+
+  for (auto tileOp : tileOps) {
+    int col = tileOp.colIndex();
+    int row = tileOp.rowIndex();
+    auto coreOp = tileOp.getCoreOp();
+    if (!coreOp) {
+      continue;
+    }
+    auto fileAttr = coreOp->getAttrOfType<StringAttr>("elf_file");
```

Comment:
Move inside the if?

Reply:
Also you can just `getElfFile()`.

And for the record (not personally @ you @spaceotter), this same thing was elsewhere in the codebase and I removed it recently. I would like to never see this again - we define attrs as `ins` params in order to exactly avoid string matching.

Reply:
Mahesh did it this way - it did occur to me that getElfFile might exist, then IntelliSense couldn't find it so I kept going...

Reply:
It indeed didn't exist until I noticed this kind of thing was happening (reliance on an arbitrarily inserted attribute) and so I added it. Like I said not your fault just that I would like to avoid anything like this in the future.